### PR TITLE
Arduino gpio fix

### LINKED ISSUE #78381: tests: gpio_basic_api: arduino not valid on different gpio controllers
**Describe the bug**
test execution **drivers.gpio.2pin_arduino**  fails in kernel panic on some stm32 boards because the overlay selects arduino connector pins on different GPIO controllers

Please also mention any information which could help others to understand
the problem you're facing:
 - example on the nucleo stm32g474re board (and  many others too)
The testcase selects arduino connector pins  8 & 9   through the arduino.overlay file
```
		out-gpios = <&arduino_header 8 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
		in-gpios = <&arduino_header 9 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
```

On some stm32 boards, like the nucleo_g474re, the arduino pins 8 & 9 are NOT on the same gpio bank 

pin 8    <8 0 &**gpioa** 10 0>,   /* D2*/
pîn 9    <9 0 &**gpiob** 3 0>,	/* D3 */

So the testcase runs into k_panic(); in the board_setup function:
```
	/* PIN_IN and PIN_OUT must be on same controller. */

	if (in_dev != out_dev) {
		printk("FATAL: output controller %s != input controller %s\n",
		       out_dev->name, in_dev->name);
		k_panic();
	}
```

 - Is this a regression? since the https://github.com/zephyrproject-rtos/zephyr/pull/77699

**To Reproduce**

**Expected behavior**
Test passed or excluded

**Impact**
CI failure

**Logs and console output**
```
FATAL: output controller gpio@48000400 != input controller gpio@48000000                                         
E: r0/a1:  0x00000004  r1/a2:  0x0000000a  r2/a3:  0x00000000                                                    
E: r3/a4:  0x00000004 r12/ip:  0x0800925f r14/lr:  0x080004d5                                                    
E:  xpsr:  0x21000000                                                                                            
E: Faulting instruction address (r15/pc): 0x080004e4                                                             
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0                                                               
E: Current thread: 0x20000228 (main)                                                                             
E: Halting system  
```

**Environment (please complete the following information):**

 - OS: ubuntu
 - Toolchain (e.g Zephyr SDK, ...) 0.16.3
 - Commit SHA or Version used :  4c6b1e5a65e752c881183ddb4687fa0231b11cc3

### PR DESCRIPTION
**Pull request by hakehuang**
st platforms require input and output on same port, add support for such constrains

#fixes: #78381

### CODE REVIEW

**Comment by erwango** [comment]
> st platforms require input and output on same port, add support for such constrains

This is wrong.

`tests/drivers/gpio/gpio_basic_api/` expects test pin 1 and test pin 2 to be on the same port.

With [your PR](https://github.com/zephyrproject-rtos/zephyr/pull/77699) pins used for the test have been moved (from A0/A1 to D2/D3) and on most STM32, D2/D3 pins are not on the same GPIO bank, hence test contraint is no more respected.

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 1]
```diff
@@ -36,20 +36,66 @@ tests:
...
+    min_flash: 34
+    depends_on:
+      - gpio
+      - arduino_gpio
+    filter: >
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and dt_res_prop_addr_cmp("in-gpios","out-gpios", 0) and CONFIG_SOC_FAMILY_STM32
+    extra_args: "DTC_OVERLAY_FILE=st_arduino.overlay"
+    extra_configs:
+      - CONFIG_ONE_PORT=y
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:76]: Default configuration, not required

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 2]
```diff
@@ -36,20 +36,66 @@ tests:
...
+    integration_platforms:
+      - stm32mp157c_dk2
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:78]: Any other one would be better (say disco_l475_iot1)

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 3]
```diff
@@ -36,20 +36,66 @@ tests:
...
+    platform_exclude:
+      # below boards are mikro bus not arduino
+      - da14695_dk_usb
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:81]: Not STM32_SOC_FAMILY

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 4]
```diff
@@ -36,20 +36,66 @@ tests:
     depends_on:
       - gpio
       - arduino_gpio
-    filter: dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+    filter: >
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and not CONFIG_SOC_FAMILY_STM32
     extra_args: "DTC_OVERLAY_FILE=arduino.overlay"
+    extra_configs:
+      - CONFIG_ONE_PORT=n
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:44]: It seems wrong that `drivers.gpio.2pin_arduino:` which looks as the default test is set with `CONFIG_ONE_PORT=n` which is not default test configuration

**Comment by hakehuang** [comment]
> With [your PR](https://github.com/zephyrproject-rtos/zephyr/pull/77699) pins used for the test have been moved (from A0/A1 to D2/D3) and on most STM32, D2/D3 pins are not on the same GPIO bank, hence test contraint is no more respected.

@erwango , thanks for the confirmation, I am modify the cases to split the port dependency.

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 5]
```diff
@@ -36,20 +36,66 @@ tests:
     depends_on:
       - gpio
       - arduino_gpio
-    filter: dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+    filter: >
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and not CONFIG_SOC_FAMILY_STM32
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:41]: ok, I see, it is not IP limitation, just case limitation, let me update

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 1 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:76]: yes

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 2 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:78]: ok

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 3 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:81]: this board declare as arduino interface in <board?.yaml but actually mikro bus interface in dts, do you know who is the maintainer for this board?

Comment (erwango) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:75]: I don't know who is maintaining it but I think it would be justified if you change it.

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 2 above)
Comment (erwango) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:72]: Thanks

File: tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c [hunk 6]
```diff
@@ -1,21 +1,22 @@
...
  */
 
-
 #include "test_gpio.h"
 
 #define ALL_BITS ((gpio_port_value_t)-1)
 
-static const struct device *const dev = DEVICE_DT_GET(DEV);
+static const struct device *const dev_in = DEVICE_DT_GET(DEV_IN);
+static const struct device *const dev_out = DEVICE_DT_GET(DEV_OUT);
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:13]: I think we should first understand why it was done this way (use `git blame` to find PR that introduced it) and if it is ok to change it.
Also, it should be done in a dedicated commit explaining all this

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c (hunk 6 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:13]: it is Peter Bigot add this original with  PR #73638 with commit #ae7a59eae4b877b1990c576b307f185c3a369484
```
test/drivers/gpio: add tests for new API

Test that the new port API functions all behave as expected, including
physical vs logical level for input and output as well as masked and
set-based output operations.  Also tests the new pin API functions.

For running on real hardware this test now uses a local test-specific
devicetree binding.  For build-only tests any platform with a GPIO
alias should be tested.

The new code increases flash requirements so add a filter to exclude
platforms that won't link.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
```

but there is not related comments

Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:13]: > Also, it should be done in a dedicated commit explaining all this

updated

**Comment by hakehuang** [comment]
@erwango i enable the CONFIG_SKIP_PULL_TEST by default, which can save some efforts on select a gpio when the board has external pull up/down

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (erwango) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:41]: So, if the test doesn't require anymore to have both pins on same GPIO bank, this "STM32 specific case" is not required anymore.

**Comment by hakehuang** [comment]
all NXP platforms with arduino header interfaces platform test PASS.

**Comment by hakehuang** [comment]
@erwango , is there any more issues found on ST platforms?

**Comment by hakehuang** [comment]
@erwango how is your test status? all NXP platform PASS.

**Comment by erwango** [comment]
> @erwango how is your test status? all NXP platform PASS.

Still on going, but I think it should be fine soon.

**Comment by hakehuang** [comment]
> For everything to work well on the STM32 side, you should rather use pins D9 and D10, so Arduino headers 15 and 16.

@erwango I update ST overlay to use D9/D10, please check. Thanks

**Comment by hakehuang** [comment]
> There were some communication issues with other tests on some series using pins D0 and D1 and a slight modification would be needed on the st_arduino.overlay.

As to this, you can check the whether you have PULL_UP on those pins, if so you need add a config to skip the PULL_TEST.

**Comment by djiatsaf-st** [comment]
> > For everything to work well on the STM32 side, you should rather use pins D9 and D10, so Arduino headers 15 and 16.
> 
> @erwango I update ST overlay to use D9/D10, please check. Thanks

I have performed hw tests on this PR and it seems correct with D9/D10

**Comment by hakehuang** [comment]
@erwango no the test pass on all NXP and ST platforms.

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 7]
```diff
@@ -36,15 +36,27 @@ tests:
...
+      - mimxrt1020_evk
+      - mimxrt1040_evk
+      - mimxrt1060_evk
+      - mimxrt1060_evkb
+      - lpcxpresso55s69/lpc55s69/cpu0
+      - lpcxpresso55s36
+      - nucleo_wb05kz
+      - nucleo_wb09ke
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:59]: Not required as filter is !CONFIG_SOC_FAMILY_STM32

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 8]
```diff
@@ -53,3 +65,30 @@ tests:
...
+      - mimxrt1020_evk
+      - mimxrt1040_evk
+      - mimxrt1060_evk
+      - mimxrt1060_evkb
+      - lpcxpresso55s69/lpc55s69/cpu0
+      - lpcxpresso55s36
+      - nucleo_wb09ke
+      - nucleo_wb05kz
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:77]: From what I understood, only wb0 boards required a specific overlay. @djiatsaf-st can you clarify the status ?

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml [hunk 9]
```diff
@@ -53,3 +65,30 @@ tests:
...
+      and CONFIG_SOC_FAMILY_STM32
+    extra_args: "DTC_OVERLAY_FILE=st_arduino.overlay"
+    integration_platforms:
+      - disco_l475_iot1
+    platform_exclude:
+      # below boards need customized
+      - nucleo_wb05kz
+      - nucleo_wb09ke
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/testcase.yaml:94]: See upper comment

File: tests/drivers/gpio/gpio_basic_api/st_arduino.overlay [hunk 10]
```diff
@@ -4,13 +4,13 @@
...
+/* loopback with arduino header D9 and D10 */
 
 / {
 	resources {
 		compatible = "test-gpio-basic-api";
 		status = "okay";
-		out-gpios = <&arduino_header 6 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
-		in-gpios = <&arduino_header 7 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
+		out-gpios = <&arduino_header 15 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
+		in-gpios = <&arduino_header 16 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
```
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/st_arduino.overlay:14]: No need for a specific commit. This should be fixed directly in the commit that introduced the file.

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/gpio/gpio_basic_api/st_arduino.overlay (hunk 10 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/st_arduino.overlay:14]: ok

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 8 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:73]: this is not a must, as it also helps to demo how to customize the test settings. agree? @erwango

**Comment by hakehuang** [comment]
the build failure is not related to this PR

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 8 above)
Comment (djiatsaf-st) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:73]: > From what I understood, only wb0 boards required a specific overlay. @djiatsaf-st can you clarify the status ?

There will be no problem for nucleo_l031k6.
For nucleo_wba55cg, the user manual specifies that we need to change state of solder bridges to use pin D8 or D9. 
So it will be better to leave as excluded.

**Comment by hakehuang** [comment]
the ci failure is not related to this PR, but #79576

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 8 above)
Comment (erwango) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml:73]: @hakehuang Please make `nucleo_l031k6` use st_arduino.overlay  and we'll be good to go

File: tests/drivers/gpio/gpio_basic_api/st_arduino.overlay
Comment (erwango) [inline tests/drivers/gpio/gpio_basic_api/st_arduino.overlay:1]: Suggestion:
I expect non STM32 boards can also use this overlay for specific constraints.
So I suggest to rename:
arduino.overlay > arduino_D0D1.overlay
arduino_st.overlay > arduino_D9D10.overlay

which will make it more generic and reuseable by others

**Review by erwango — CHANGES_REQUESTED** [review]
Last 2 comments

**Review by erwango — APPROVED** [review]
Thanks for the changes

File: tests/drivers/gpio/gpio_basic_api/testcase.yaml (hunk 8 above)
Comment (hakehuang) [reply tests/drivers/gpio/gpio_basic_api/testcase.yaml]: done

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by erwango** [comment]
@nashif Please have a look

**Review by nashif — APPROVED** [review]

**Review by nashif — APPROVED** [review]
